### PR TITLE
Code cleanup for future events - [MOD-5646]

### PR DESCRIPTION
**Pull request by GuyAv46**
**Describe the changes in the pull request**

Some code cleanup related to [MOD-5646](https://redislabs.atlassian.net/browse/MOD-5646). This PR removes code duplications and unnecessary casting, in an attempt to make future bugs easier to fix, and their stack traces easier to understand.

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

[MOD-5646]: https://redislabs.atlassian.net/browse/MOD-5646?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/inverted_index.c [hunk 1]
```diff
@@ -120,6 +119,14 @@ void IndexReader_OnReopen(void *privdata) {
     }
   }
 
+  IndexReader_OnReopen(ir);
+}
+
+void IndexReader_OnReopen(IndexReader *ir) {
```
Comment (ephraimfeldblum) [inline src/inverted_index.c:125]: good splitting out shared code to its own function. removes unnecessary duplication.

File: src/inverted_index.c [hunk 2]
```diff
@@ -130,12 +137,10 @@ void IndexReader_OnReopen(void *privdata) {
...
 
-    // reset the state of the reader
+    // keep the last docId we were at
     t_docId lastId = ir->lastId;
-    ir->currentBlock = 0;
-    ir->br = NewBufferReader(&IR_CURRENT_BLOCK(ir).buf);
-    ir->lastId = IR_CURRENT_BLOCK(ir).firstId;
-
+    // reset the state of the reader
+    IR_Rewind(ir);
```
Comment (ephraimfeldblum) [inline src/inverted_index.c:143]: good removal of duplicate code.

File: src/numeric_index.c [hunk 3]
```diff
@@ -855,31 +855,6 @@ void NumericRangeTreeIterator_Free(NumericRangeTreeIterator *iter) {
   rm_free(iter);
 }
 
-
-void IndexReader_Reopen(IndexReader *ir, void *privdata) {
```
Comment (ephraimfeldblum) [inline src/numeric_index.c:859]: removal of unnecessary duplicate.

File: src/index.c [hunk 4]
```diff
@@ -227,11 +227,11 @@ IndexIterator *NewUnionIterator(IndexIterator **its, int num, DocTable *dt, int
   return it;
 }
 
-void UI_Foreach(IndexIterator *index_it, void (*callback)(IndexReader *it, void *privdata), void *privdata) {
+void UI_Foreach(IndexIterator *index_it, void (*callback)(IndexReader *it)) {
```
Comment (ephraimfeldblum) [inline src/index.c:230]: `privdata` was not being used at all.

File: src/tag_index.c [hunk 5]
```diff
@@ -192,27 +192,8 @@ static void TagReader_OnReopen(void *privdata) {
         return;
       }
     }
-
-    // the gc marker tells us if there is a chance the keys has undergone GC while we were asleep
```
Comment (ephraimfeldblum) [inline src/tag_index.c:196]: removal of unnecessary duplicate.

**Review by ephraimfeldblum — APPROVED** [review]
Is it prudent to add a test here?
This was almost all a refactor to remove duplicate code.

File: src/numeric_index.c [hunk 6]
```diff
@@ -899,8 +874,8 @@ void NumericRangeIterator_OnReopen(void *privdata) {
   }
 
   if (it->type == READ_ITERATOR) {
-    IndexReader_Reopen(it->ctx, nu);
+    IndexReader_OnReopen(it->ctx);
   } else if (it->type == UNION_ITERATOR) {
-    UI_Foreach(it, IndexReader_Reopen, nu);
+    UI_Foreach(it, IndexReader_OnReopen);
```
Comment (alonre24) [inline src/numeric_index.c:879]: A union of unions is not possible? I recall that you mentioned something about this case being not taken care of

Comment (GuyAv46) [reply src/numeric_index.c:879]: On geo queries there is 0-9 options to check (related to the hashing method) and in the “worst” case we have a union of 9 numeric ranges, each one could be a union of index readers. But each numeric range is registered separately to “onReopen” so here there is no other possibility. We could add here a log assert if it’s neither of these cases

Comment (alonre24) [reply src/numeric_index.c:879]: then lets do it

**Review by alonre24 — COMMENTED** [review]
Try checking the possibility that union iterator was corrupted (number of children is not the number we think), and maybe add more info for logs in case we crash again

**Review by alonre24 — DISMISSED** [review]

**Review by alonre24 — APPROVED** [review]

**Pull request merged**